src/middlewared/middlewared/plugins/pool.py

```diff
@@ -2112,6 +2099,16 @@ async def lock(self, id, options):
         elif id == (await self.middleware.call('systemdataset.config'))['pool']:
             raise CallError(f'Please move system dataset to another pool before locking {id}')
 
+        async def detach(delegate):
+            attachments = await delegate.query(self.__attachments_path(ds), True, {'locked': False})
+            if isinstance(delegate, LockableFSAttachmentDelegate):
+                await delegate.detach(attachments)
+            else:
+                await delegate.toggle(attachments, False)
```

Comment:
Which services will cause trouble locking the dataset if we don't disable them? How do we plan on re-enabling them when unlocking the dataset?

Reply:
So we have 4 attachment delegates which do not use the lockable fs attachment delegate class. For 2 of them, locking/unlocking is a no op ( replication / snapshot ), rest are jails and vms. For the latter 2, toggle should be enough and if they are unlocked they should start as desired.

Reply:
I think we need `stop` method in `AttachmentDelegate`:
* Default implementation is `raise NotImplementedError`
* for replication/snapshot it will be noop (we'll probably need to learn to terminate replication task someday though)
* for jail/vm it will be stopping jail/vm
* for `LockableFSAttachmentDelegate` it will be noop
* for Samba it will be `smbcontrol smbd close-share`

That will:

* Solve these ugly `isinstance` call issue
* Prevent attachments from accidentally being disabled
* Get rid of arguments for `restart_reload_services`